### DISCUSSION TITLE
Comparison to GunDB?

### DISCUSSION THREAD

**Discussion by redbar0n** [tantaman/aphrodite#30] [Q&A] [answered]
How does Aphrodite compare and constrast with [GunDB](https://github.com/amark/gun)?

**Comment by tantaman** [comment, marked as answer]
Gun is definitely much more mature and feature complete. On Aphrodite, I'm still working towards implementing p2p capabilities and permission policy evaluation and there is much more work to do to make all of this production ready.

All that aside, however, I think the difference is similar to the difference between dynamic and statically typed languages.

Aphrodite requires you to define a schema for your data model so we know, up front, the shape that your data can take.
Gun has no schema.

I went the schema route for a few reasons:

1) Having a schema lets us understand when peers are on different versions and allows us to provide support for migrating one peer's data format to another peer's. There's some interesting research on the topic here: https://www.inkandswitch.com/cambria/
2) The schema allows us to generate strongly typed query builders, mutators and models to help prevent mistakes. Example of a generated query builder below

https://user-images.githubusercontent.com/1009003/175358998-ac596fc8-ac24-4cc8-ac25-795c41995192.mov

3) Permissions can be defined [statically on the schema](https://github.com/tantaman/aphrodite/tree/main/extensions/authorization-grammar), allowing you to more easily reason about who can see what. This is similar to [Postgres row level security](https://www.postgresql.org/docs/current/ddl-rowsecurity.html) or [Oso](https://www.osohq.com/) if you're familiar with those.
4) You'll be able to choose different consistency rules per-schema. I.e., some nodes in your system might be simple "last write wins", others might need something more complex. I'm not aware if this is possible in GunDB.


The other difference that sticks out to me is that Aphrodite builds atop sqlite (with the option of using postgres instead) rather than re-implementing the database layer. Other research on this topic -- https://hal.inria.fr/hal-02983557/document & https://munin.uit.no/bitstream/handle/10037/22344/thesis.pdf?sequence=2

If you want to get more of a feel for it, I put up an interactive notebook here: https://observablehq.com/@tantaman/aphrodite-chinook
